docs/src/main/asciidoc/security.adoc

```diff
@@ -96,12 +98,79 @@ Additionally it provides `JWT Generation API` for creating `signed`, `inner-sign
 
-=== OAuth2
+==== OAuth2
 
 `quarkus-elytron-security-oauth2` provides an alternative to `quarkus-oidc` Bearer Token Authentication Mechanism. It is based on `Elytron` and is primarily meant for introspecting the opaque tokens remotely.
 
 See the link:security-oauth2[Using OAuth2] guide for more information.
 
+==== Choosing between OpenId Connect, SmallRye JWT and OAuth2 extensions
+
+`quarkus-oidc` extension requires an OpenId Connect provider such as Keycloak which can be used to verify the Bearer tokens or authenticate the human users with the Authorization Code flow. In both cases `quarkus-oidc` requires a connection to this OpenId Connect provider.
+
+`quarkus-oidc` is the only option when the user authentication via Authorization Code flow or supporting multiple tenants is required. It can also request a UserInfo using both Authorization Code Flow and Bearer access tokens.
+
+When the Bearer tokens have to be verified then `quarkus-oidc`, `quarkus-smallrye-jwt` and `quarkus-elytron-security-oauth2` can be used.
+
+If you have Bearer tokens in a JWT format then all these 3 extensions can be used. Both `quarkus-oidc` and `quarkus-smallrye-jwt` support refreshing the JsonWebKey (JWK) set when the OpenId Connect provider rotates the keys, therefore `quarkus-oidc` or `quarkus-smallrye-jwt` should be used for verifying JWT tokens if the remote token introspection has to be avoided or not supported by the providers.
+
+`quarkus-smallrye-jwt` does not support the remote introspection of the opaque tokens or even JWT tokens - it always relies on the locally available keys - possibly fetched from the OpenId Connect provider. So if you need to introspect the JWT tokens remotely then both `quarkus-oidc` and `quarkus-elytron-security-oauth2` will work. Both extensions also support the verification of the opaque/binary tokens via the remote introspection.
+
+`quarkus-oidc` and `quarkus-smallrye-jwt` can have both JWT and opaque tokens injected into the endpoint code - the injected JWT tokens may offer a richer information about the user. All extensions can have the tokens injected as `Principal`.
+
+`quarkus-smallrye-jwt` supports more key formats than `quarkus-oidc`. The latter will only use the JWK-formatted keys which are part of a JWK set. The former - can also work with PEM keys.
+
+`quarkus-smallrye-jwt` can handle locally not only signed but also inner-signed-and-encrypted or only encrypted tokens. In fact `quarkus-oidc` and `quarkus-elytron-security-oauth2` can verify such tokens too but only by treating them as opaque tokens and verifying them via the remote introspection.
+
+`quarkus-elytron-security-oauth2` is the best choice if you need a light weight library for the remote introspection of either JWT or opaque tokens.
+
+Below is a summarizing the options.
+
+|===
+| | quarkus-oidc| quarkus-smallrye-jwt | quarkus-elytron-security-oauth2
+
+|Bearer JWT verification is required
+|Yes
+|Yes
+|Yes
```

Comment:
Are you sure that the JWT will be verified by OAuth2 ? I think that the token will be validated by the introspection endpoint so it is the introspection endpoint responsability to verify the token in this case

Reply:
@loicmathieu One can use this extension, in principle, to point it to Keycloak introspection endpoint and introspect everything - but I see your point, I think I'll clarify it - that it depends on whether the server recognizes JWT or not (JWT are used in the OAuth2 space but yeah, it depends)

Reply:
@loicmathieu @gastaldi  sorry, I misunderstood - yeah - I've qualified it differently (`Introspection` vs `yes`) in this case.

Also added a short note about the opaque vs JWT token formats from the architectural point of view

have a look please